identity/aziot-identity-common/src/lib.rs

```diff
@@ -70,81 +70,133 @@ pub struct IoTHubDevice {
-        pub symmetric_key: Option<SymmetricKey>,
-
-        #[serde(skip_serializing_if = "Option::is_none")]
-        pub x509_thumbprint: Option<X509Thumbprint>,
-
-        #[serde(skip_serializing_if = "Option::is_none")]
-        pub type_: Option<AuthType>,
-    }
-
-    #[derive(Clone, Debug, serde::Serialize, serde::Deserialize, PartialEq)]
-    #[serde(rename_all = "camelCase")]
-    pub struct Module {
-        pub module_id: String,
-        #[serde(skip_serializing_if = "Option::is_none")]
-        pub managed_by: Option<String>,
-
-        pub device_id: String,
-        #[serde(skip_serializing_if = "Option::is_none")]
-        pub generation_id: Option<String>,
-        #[serde(skip_serializing_if = "Option::is_none")]
-        pub authentication: Option<AuthMechanism>,
-    }
+	#[derive(Clone, Copy, Debug, serde::Deserialize, PartialEq, serde::Serialize)]
+	#[serde(rename_all = "camelCase")]
+	pub enum AuthType {
+		None,
+		Sas,
+		X509,
+	 }
+
+	#[derive(Clone, Debug, serde::Serialize, serde::Deserialize, PartialEq)]
+	#[serde(rename_all = "camelCase")]
+	pub struct X509Thumbprint {
+		#[serde(skip_serializing_if = "Option::is_none")]
+		pub primary_thumbprint: Option<String>,
+		#[serde(skip_serializing_if = "Option::is_none")]
+		pub secondary_thumbprint: Option<String>,
+	}
+
+	#[derive(Clone, Debug, serde::Serialize, serde::Deserialize, PartialEq)]
+	#[serde(rename_all = "camelCase")]
+	pub struct SymmetricKey {
+		#[serde(skip_serializing_if = "Option::is_none")]
+		pub primary_key: Option<http_common::ByteString>,
+		#[serde(skip_serializing_if = "Option::is_none")]
+		pub secondary_key: Option<http_common::ByteString>,
+	}
+
+	#[derive(Clone, Debug, serde::Serialize, serde::Deserialize, PartialEq)]
+	#[serde(rename_all = "camelCase")]
+	pub struct AuthMechanism {
+		#[serde(skip_serializing_if = "Option::is_none")]
+		pub symmetric_key: Option<SymmetricKey>,
+
+		#[serde(skip_serializing_if = "Option::is_none")]
+		pub x509_thumbprint: Option<X509Thumbprint>,
+
+		#[serde(skip_serializing_if = "Option::is_none")]
+		pub type_: Option<AuthType>,
+	}
+
+	#[derive(Clone, Debug, serde::Serialize, serde::Deserialize, PartialEq)]
+	#[serde(rename_all = "camelCase")]
+	pub struct Module {
+		pub module_id: String,
+		#[serde(skip_serializing_if = "Option::is_none")]
+		pub managed_by: Option<String>,
+
+		pub device_id: String,
+		#[serde(skip_serializing_if = "Option::is_none")]
+		pub generation_id: Option<String>,
+		#[serde(skip_serializing_if = "Option::is_none")]
+		pub authentication: Option<AuthMechanism>,
+	}
+}
+
+/// Common trait for identity clients.
+#[async_trait::async_trait]
+pub trait IdentityClient {
+	/// Type of identity, i.e. Hub or local.
+	type IdentityType;
+
+	/// Device context for constructor, unique per identity type.
+	type DeviceInfo;
+
+	/// Context for creating and updating identities, unique per identity type.
+	type IdentityInfo;
+
+	/// Constructor: create a new identity client.
+	fn new(
```

Comment:
This is an implementation specific constructors, which shouldn't be part of a trait.